crates/codegen/ebnf/src/builder.rs

```diff
@@ -0,0 +1,432 @@
+
+                values.push(Value::new(Self::build_ref(base_name), None));
+            }
+            OperatorModel::Postfix => {
+                leading_comments.push("Postfix unary operator".to_string());
+
+                values.insert(0, Value::new(Self::build_ref(base_name), None));
+            }
+            OperatorModel::BinaryLeftAssociative => {
+                leading_comments.push("Left-associative binary operator".to_string());
+
+                values.insert(0, Value::new(Self::build_ref(base_name), None));
+                values.push(Value::new(Self::build_ref(base_name), None));
+            }
+            OperatorModel::BinaryRightAssociative => {
+                leading_comments.push("Right-associative binary operator".to_string());
+
+                values.insert(0, Value::new(Self::build_ref(base_name), None));
+                values.push(Value::new(Self::build_ref(base_name), None));
+            }
+        };
+
+        leading_comments.extend(Self::build_enabled(enabled));
+
+        self.add_definition(name, leading_comments, values, DefinitionKind::Sequence);
+    }
+
+    fn build_fields(fields: &IndexMap<Identifier, Field>) -> Vec<Value> {
+        fields
+            .values()
+            .map(|field| match field {
+                Field::Required { reference } => Value::new(Self::build_ref(reference), None),
+                Field::Optional { reference, enabled } => Value::new(
+                    Expression::new_optional(Self::build_ref(reference).into()),
+                    Self::build_enabled(enabled),
+                ),
+            })
+            .collect()
+    }
+
+    fn build_trivia_item(&mut self, trivia_item: &TriviaItem) {
+        let TriviaItem { name, scanner } = trivia_item;
+
+        self.add_entry(name, true, false);
+
+        self.add_definition(
+            name,
+            None,
+            Some(Value::new(Self::build_scanner(scanner), None)),
+            DefinitionKind::Sequence,
+        );
+    }
+
+    fn build_keyword_item(&mut self, keyword_item: &KeywordItem) {
+        let KeywordItem {
+            name,
+            identifier: _,
+            definitions,
+        } = keyword_item;
+
+        self.add_entry(name, true, false);
+
+        for KeywordDefinition {
+            enabled,
+            reserved,
+            value,
+        } in definitions
+        {
+            let mut leading_comments = vec![];
+
+            leading_comments.extend(Self::build_enabled(enabled));
+            leading_comments.extend(Self::build_reserved(reserved));
+
+            self.add_definition(
+                name,
+                leading_comments,
+                Some(Value::new(Self::build_keyword_value(value), None)),
+                DefinitionKind::Sequence,
+            );
+        }
+    }
+
+    fn build_token_item(&mut self, token_item: &TokenItem) {
+        let TokenItem { name, definitions } = token_item;
+
+        self.add_entry(name, true, false);
```

Comment:
multiple bare boolean arguments are hard to follow when reading multiple of these functions (wait, was the first one terminal or inlined?).

Simplest idea is to add simple two-state enums and use them; other than that, it makes sense to encapsulate that logic in a trait that allows you to create an EBNF ID from the named `XYZItem` variant. This would have the added benefit of encoding the fact, that `add_entry` is only really callable with XYZItem definitions (groups calls, facilitates reasoning), rather than a simple (Identifier, bool, bool) tuple.

Also, since the pattern since to be `add_entry(..);` followed by `self.add_definition();` which only pushes to a vec inside a given map entry, it might make sense to see if we can create a version of `add_entry` that returns the entry and allows you to directly append `Definition`s for simplicity but that's well into nice-to-have area; just thinking out loud.

Reply:
I agree as well if this was a publicly used method, but having a trait and 9-10 implementations for a private helper used within one type is too much work/too much complex for the benefit it adds IMO.

Also, `add_entry()` and `add_definition()` are not always called together, as many items add one entry and multiple definitions. So I think the uniformity here is worth it.

Reply:
I understand that the above might be overengineering. In which case, introducing trivial enums like `Terminal::Yes`, `Inline::No` would be a great and simple alternative :+1: 

Reading
```rust
self.add_entry(name, Terminal::Yes, Inlined::No)
```
rather than
```rust
self.add_entry(name, true, false)
```

is a lot more readable and is easier to understand what's going on, I'm sure you agree as well :crossed_fingers:

Reply:
Sure